analyzers/src/SonarAnalyzer.Common/Rules/IndexOfCheckAgainstZeroBase.cs

```diff
@@ -0,0 +1,103 @@
+                "LastIndexOf",
+                "LastIndexOfAny"
+            };
+
+        private protected static readonly ImmutableArray<KnownType> CheckedTypes =
+            ImmutableArray.Create(
+                KnownType.System_Array,
+                KnownType.System_Collections_Generic_IList_T,
+                KnownType.System_String,
+                KnownType.System_Collections_IList
+            );
+
+        private readonly DiagnosticDescriptor rule;
+
+        protected abstract ILanguageFacade LanguageFacade { get; }
+        protected abstract TSyntaxKind LessThanExpression { get; }
+        protected abstract TSyntaxKind GreaterThanExpression { get; }
+
+        protected abstract SyntaxNode Left(TBinaryExpressionSyntax binaryExpression);
+        protected abstract SyntaxNode Right(TBinaryExpressionSyntax binaryExpression);
+        protected abstract SyntaxToken OperatorToken(TBinaryExpressionSyntax binaryExpression);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected IndexOfCheckAgainstZeroBase(System.Resources.ResourceManager rspecResources) =>
+            rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, rspecResources);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                LanguageFacade.GeneratedCodeRecognizer,
+                c =>
+                {
+                    var lessThan = (TBinaryExpressionSyntax)c.Node;
+                    if (IsInvalidComparison(Left(lessThan), Right(lessThan), c.SemanticModel))
+                    {
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, Left(lessThan).CreateLocation(OperatorToken(lessThan))));
+                    }
+                },
+                LessThanExpression);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                LanguageFacade.GeneratedCodeRecognizer,
+                c =>
+                {
+                    var greaterThan = (TBinaryExpressionSyntax)c.Node;
+                    if (IsInvalidComparison(Right(greaterThan), Left(greaterThan), c.SemanticModel))
+                    {
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, OperatorToken(greaterThan).CreateLocation(Right(greaterThan))));
+                    }
+                },
+                GreaterThanExpression);
+        }
+
+        private bool IsInvalidComparison(SyntaxNode constantExpression, SyntaxNode methodInvocationExpression, SemanticModel semanticModel) =>
+            LanguageFacade.ExpressionNumericConverter.TryGetConstantIntValue(constantExpression, out var constValue)
+            && constValue == 0
+            && IsSensitiveCall(methodInvocationExpression, semanticModel);
+
+        private bool IsSensitiveCall(SyntaxNode call, SemanticModel semanticModel) =>
```

Comment:
_To consider:_ I'm not sure if the name `IsSensitiveCall` adds enough value to deserve 3 lines of code. Replacing it with `&&` to do longer condition would work just as well.
```suggestion
            && 
```